File: UnrealAgent.Frontend/UI/Input/ChatInput.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using UnrealAgent.Backend.Conversation;
using UnrealAgent.Frontend.Infrastructure;

namespace UnrealAgent.Frontend.UI.Input;


public partial class ChatInput :JsComponentBase
{
    // 메세지 전송 콜백
    [Parameter] public EventCallback<UserInput> OnSend { get; set; }
    
    // textarea 요소 참조
    private ElementReference TextAreaRef;
    
    // .Net에서 JS 가 호출할 수 있는 참조
    private DotNetObjectReference<ChatInput>? DotNetRef;
    
    // 모드 스위치 컴포넌트 참조
    private ModeSwitcher ModeSwitcherRef = null!;
    
    // 커맨드 팝업 참조
    private CommandPopup CmdPopup = null!;
    
    // 멘션 팝업 참조
    private MentionPopup MenPopup = null!;
    
    // textarea 바인딩 값, 변경 시 커맨드 팝업을 갱신
    // 현재 입력 텍스트
    private string InputText
    {
        get;
        set
        {
            field = value;
            CmdPopup.Update(value);
            MenPopup.Update(value);
        }
    } = "";
    
    // [+] 패널(모델/Thinking/Effort/모드)이 펼쳐져 있는지 여부
    private bool bShowPanel;

    // 전송 버튼을 노란색으로 활성화할지 판단한다
    private bool bHasText => !string.IsNullOrWhiteSpace(InputText);
    
    // [+] 버튼 : 패널을 여닫는다
    private void TogglePanel() => bShowPanel = !bShowPanel;


    // JS 모듈 로드 후 Enter 키 바인딩 설정
    protected override async Task OnModuleLoaded()
    {
        DotNetRef = DotNetObjectReference.Create(this);
        await Module.InvokeVoidAsync("setupKeyBindings", TextAreaRef, DotNetRef);
        // 입력 내용에 따라 textarea 높이를 늘려 준다 (카톡과 동일한 동작)
        await Module.InvokeVoidAsync("setupAutoGrow", TextAreaRef);
    }
    
    // Shift + Tab 시 JS에서 호출된다
    [JSInvokable]
    public void CycleMode() => ModeSwitcherRef.CycleMode();
    
    // 팝업에서 방향키로 항목을 탐색
    [JSInvokable]
    public async Task PopupNavigate(int Direction) => await CmdPopup.Navigate(Direction);

    // 팝업에서 현재 선택된 항목을 적용한다
    [JSInvokable]
    public void PopupSelect()
    {
        string? Result = CmdPopup.Select();

        if (Result is not null)
        {
            InputText = Result;
            StateHasChanged();
        }
    }
    
    
    // 팝업을 닫는다
    [JSInvokable]
    public void PopupClose() => CmdPopup.Close();
    
    
    ///////////////////////////////////////////////////////////////////////////
    ///  멘션 팝업 
    ///////////////////////////////////////////////////////////////////////////
    
    // 멘션 팝업에서 방향키로 항목을 탐색한다.
    [JSInvokable]
    public async Task MentionNavigate(int Direction) => await MenPopup.Navigate(Direction);
    
    // 멘션 팝업에서 Enter로 최종 선택
    [JSInvokable]
    public void MentionSelect()
    {
        string? Path = MenPopup.Select();
        if (Path is null)
            return;

        int AtIndex = InputText.LastIndexOf('@');
        if (AtIndex < 0)
            return;

        InputText = InputText[..AtIndex] + Path + " ";
        
        StateHasChanged();
    }
    
    /// <summary>멘션 팝업에서 Tab으로 드릴다운/선택합니다.</summary>
    [JSInvokable]
    public void MentionTab()
    {
        string? Result = MenPopup.Tab();
        if (Result is null) 
            return;

        int AtIndex = InputText.LastIndexOf('@');
        if (AtIndex < 0) 
            return;

        InputText = Result.EndsWith('/')
            ? InputText[..AtIndex] + "@" + Result
            : InputText[..AtIndex] + Result + " ";
        
        StateHasChanged();
    }
    
    /// <summary>멘션 팝업에서 ← 키로 상위 이동합니다.</summary>
    [JSInvokable]
    public void MentionGoBack()
    {
        string? Result = MenPopup.GoBack();
        if (Result is null) 
            return;

        int AtIndex = InputText.LastIndexOf('@');
        if (AtIndex < 0) 
            return;

        InputText = InputText[..AtIndex] + "@" + Result;
        StateHasChanged();
    }

    /// <summary>멘션 팝업을 닫습니다.</summary>
    [JSInvokable]
    public void MentionClose() => MenPopup.Close();

    /// <summary>폼 제출 시 메시지를 전송합니다.</summary>
    private async Task HandleSubmit()
    {
        string Trimmed = InputText.Trim();

        if (string.IsNullOrEmpty(Trimmed))
            return;

        InputText = "";
        // 늘어나 있던 textarea 높이를 한 줄로 되돌린다.
        // Module은 첫 렌더 이후에만 채워지므로 null 검사를 함께 한다.
        if (Module is not null)
            await Module.InvokeVoidAsync("resetHeight", TextAreaRef);

        await OnSend.InvokeAsync(Trimmed);    }
    
    
}

File: UnrealAgent.Frontend/UI/Input/ChatInput.razor.js

// textarea에 키 바인딩을 설정합니다.
// 멘션 팝업 열림 시: ArrowUp/Down → 탐색, Enter → 선택, Tab → 드릴다운, ArrowLeft → 상위, Escape → 닫기.
// 커맨드 팝업 열림 시: ArrowUp/Down → 탐색, Enter/Tab → 선택, Escape → 닫기.
export function setupKeyBindings(textarea, dotNetRef) 
{
    textarea.addEventListener("keydown", function (e)
    {

        // 멘션 팝업이 열려있을 때
        if (document.querySelector(".mention-popup"))
        {
            switch (e.key)
            {
                case "ArrowUp":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("MentionNavigate", -1);
                    return;
                case "ArrowDown":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("MentionNavigate", 1);
                    return;
                case "Enter":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("MentionSelect");
                    return;
                case "Tab":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("MentionTab");
                    return;
                case "ArrowLeft":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("MentionGoBack");
                    return;
                case "Escape":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("MentionClose");
                    return;
            }
        }
        
        // 팝업이 열려져있을 때 키 처리
        if(document.querySelector(".command-popup"))
        {
            switch (e.key)
            {
                case "ArrowUp":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("PopupNavigate", -1);
                    return;
                case "ArrowDown":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("PopupNavigate", 1);
                    return;
                case "Tab":
                case "Enter":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("PopupSelect");
                    return;
                case "Escape":
                    e.preventDefault();
                    dotNetRef.invokeMethodAsync("PopupClose");
                    return;
            }
        }
        
        // 기본 키 바인딩
        if(e.key === "Enter" && !e.shiftKey)
        {
            e.preventDefault();
            textarea.closest("form").requestSubmit();
        }
        else if(e.key === "Tab" && e.shiftKey)
        {
            e.preventDefault();
            dotNetRef.invokeMethodAsync("CycleMode");
        }
    });
}

// 입력 내용에 맞춰 textarea 높이를 늘린다 (카카오톡처럼 한 줄로 시작해 점점 커짐)
// 높이 상한은 CSS의 max-h-[160px]가 잡아 주고, 넘치면 내부 스크롤이 생긴다.
export function setupAutoGrow(textarea)
{
    const grow = function ()
    {
        // 먼저 auto로 되돌려야 scrollHeight가 '현재 높이'가 아닌 '내용에 필요한 높이'를 준다
        textarea.style.height = "auto";
        textarea.style.height = textarea.scrollHeight + "px";
    };

    textarea.addEventListener("input", grow);
    grow();
}

// 전송 후 높이를 한 줄로 되돌린다
export function resetHeight(textarea)
{
    textarea.style.height = "auto";
}